File: 62_Unique_Paths.py
# Time Complexity : O(m * n) [given input]
# Space Complexity : O(m*n) [given input]
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        if m ==n and n ==1:
            return 1
        paths = [[0 for col in range(n)]  for row in range(m)]
        
        for i in range(1, n):
            paths[0][i] = 1
            
        for i in range(1, m):
            paths[i][0] = 1
            
        for i in range(1, m):
            for j in range(1, n):
                paths[i][j] = paths[i][j-1] + paths[i-1][j]
        
        return paths[m-1][n-1]

File: 62_unique_paths.java
    /*  Explanation
    # Leetcode problem link : https://leetcode.com/problems/unique-paths/
    Time Complexity for operators : o(n*m) .. n is the length of the string
    Extra Space Complexity for operators : o(n*m) .. n for dp array and m for hashset
    Did this code successfully run on Leetcode : NA
    Any problem you faced while coding this : No
# Your code here along with comments explaining your approach
        # Basic approach : Two for loops and find the minimum window that has string. o(n*2) time complexity
        # Optimized approach: 
                              
            # 1. 
                    A) Create An DP matrix with 1 added in first row and first column.
                    B) Then, two for loops and create the matrix. by summing upper and left column.
                    C) At the end, return last block.
    */ 

class Solution {
    public int uniquePaths(int m, int n) {
        
        // time limit exceeded for this
    //         if (m == 1 || n == 1) {
    //   return 1;
    // }
    // return uniquePaths(m - 1, n) + uniquePaths(m, n - 1);
        
        
        // DP solution for it
        
        int dp[][] = new int[m][n];
        
        for(int i=0;i<m;i++){
            dp[i][0] = 1;
        }
        
        for(int i=0;i<n;i++){
            dp[0][i] = 1;
        }
        
        for(int i=1;i<m;i++){
            for(int j=1;j<n;j++){
                dp[i][j] = dp[i-1][j] + dp[i][j-1];
            }
        }
        
        return dp[m-1][n-1];
    }
}